meshio/gmsh/_gmsh22.py

```diff
@@ -352,10 +352,10 @@ def _write_elements(fh, cells, tag_data, binary):
     for k, (cell_type, node_idcs) in enumerate(cells):
         tags = []
         for name in ["gmsh:physical", "gmsh:geometrical", "cell_tags"]:
-            try:
+            if name in tag_data:
                 tags.append(tag_data[name][k])
-            except KeyError:
-                pass
+            elif name in ["gmsh:physical", "gmsh:geometrical"]:
```

Comment:
So if physical or geometrical tags are not present, write zeros? I'm really not sure about this. If the user wants those tags written out, she can manually add them, right?

Reply:
That's basically what gmsh does. The physical and geometrical tags are always written in msh22 format. See also the number-of-tags entry [here](http://gmsh.info/doc/texinfo/gmsh.html#MSH-file-format-version-2-_0028Legacy_0029). A zero means no tag and is used to ensure the correct order if the physical and/or geometrical tag is missing.

Reply:
When you write "here", it's not clear what you mean. Please be more specific.

Reply:
I mean this section of the gmsh documentation (following that link):

> number-of-tags gives the number of integer tags that follow for the n-th element. By default, the first tag is the tag of the physical entity to which the element belongs; the second is the tag of the elementary model entity to which the element belongs; the third is the number of mesh partitions to which the element belongs, followed by the partition ids (negative partition ids indicate ghost cells). A zero tag is equivalent to no tag. Gmsh and most codes using the MSH 2 format require at least the first two tags (physical and elementary tags).

Reply:
Okay, good. If one of the two "default" tags is missing, I think we should issue warning. I don't care much if we add the data (zeros) and warn about it or don't add it and show a different warning. This should happen at the very beginning of the write step.